### PR TITLE
Add TimeBasedSizeOptimizingRotationAndRetentionTest

### PR DESCRIPTION
**Pull request by mpfz0r**
This test creates indices, forwards the clock,
calls the rotations and retentions and asserts that the list of open indices gets modified accordingly.

Use the JobSchedulerClock which can be used forward the time. Extend the JobSchedulerClock to support java.time Instant and ZonedDateTime.

/jenkins-pr-deps https://github.com/Graylog2/graylog-plugin-enterprise/pull/4572

/nocl

### CODE REVIEW

File: graylog2-server/src/main/java/org/graylog2/indexer/retention/strategies/AbstractIndexRetentionStrategy.java [hunk 1]
```diff
@@ -70,14 +72,15 @@ private void retainTimeBased(IndexSet indexSet, TimeBasedSizeOptimizingStrategyC
...
+        final long cutoff = clock.instantNow().minus(smartConfig.indexLifetimeSoft()).toEpochMilli();
         final int removeCount = (int)deflectorIndices.keySet()
                 .stream()
                 .filter(indexName -> !indices.isReopened(indexName))
+                .filter(indexName -> !hasCurrentWriteAlias(indexSet, deflectorIndices, indexName))
                 .filter(indexName -> {
                     DateTime closingDate = indices.indexClosingDate(indexName)
                             .orElseThrow(() -> new IllegalStateException(f("Index %s has no closing date - retention failed", indexName)));
-                    return closingDate.isBefore(cutoff);
+                    return closingDate.isBefore(cutoff + 1);
```
Comment (patrickmann) [inline graylog2-server/src/main/java/org/graylog2/indexer/retention/strategies/AbstractIndexRetentionStrategy.java:83]: 1 ms change - what difference does that make when retention periods are measured in days?

Comment (mpfz0r) [reply graylog2-server/src/main/java/org/graylog2/indexer/retention/strategies/AbstractIndexRetentionStrategy.java:83]: It make the comparison inclusive instead of exclusive. doesn't matter that much, but it felt nicer while testing it.

**Comment by mpfz0r** [comment]
`SuggestionResourceIT.testSuggestionsAreLimitedToStream()` test failure. has nothing to do with my change. ignoring it :shrug:

File: graylog2-server/src/main/java/org/graylog2/indexer/rotation/strategies/TimeBasedSizeOptimizingStrategy.java [hunk 2]
```diff
@@ -100,21 +105,25 @@ protected Result shouldRotate(final String index, IndexSet indexSet) {
...
 
-        return creationDate.isBefore(rotationLimit.toEpochMilli());
+        final Duration between = Duration.between(Instant.ofEpochMilli(creationDate.getMillis()), now);
+        return between.compareTo(Duration.ofDays(ROTATION_PERIOD.getDays())) >= 0;
+        //return creationDate.isBefore(rotationLimit.toEpochMilli());
     }
 
     private boolean indexExceedsSizeLimit(long size) {
-        return size > MAX_INDEX_SIZE.toBytes();
+        return size >= MAX_INDEX_SIZE.toBytes();
```
Comment (patrickmann) [inline graylog2-server/src/main/java/org/graylog2/indexer/rotation/strategies/TimeBasedSizeOptimizingStrategy.java:126]: The name `indexExceedsSizeLimit` is a bit misleading, if we allow equality. Also the log messages that rely on this method explicitly say "exceeded".
1 byte up or down is insignificant for what is typically measured in GB. Why is this change needed?

Comment (mpfz0r) [reply graylog2-server/src/main/java/org/graylog2/indexer/rotation/strategies/TimeBasedSizeOptimizingStrategy.java:126]: to accommodate the test. but yeah, I can revert that

File: graylog2-server/src/main/java/org/graylog2/indexer/rotation/strategies/TimeBasedSizeOptimizingStrategy.java [hunk 3]
```diff
@@ -100,21 +105,25 @@ protected Result shouldRotate(final String index, IndexSet indexSet) {
...
 
     private boolean indexExceedsLeeWay(DateTime creationDate, Period leeWay) {
-        final Instant now = Instant.now();
+        final Instant now = clock.instantNow();
         final Instant leewayLimit = now.minus(ROTATION_PERIOD.plus(leeWay));
 
-        return creationDate.isBefore(leewayLimit.toEpochMilli());
+        final Duration between = Duration.between(Instant.ofEpochMilli(creationDate.getMillis()), now);
+        return between.compareTo(Duration.ofDays(ROTATION_PERIOD.plus(leeWay).getDays())) >= 0;
+        //return creationDate.isBefore(leewayLimit.toEpochMilli());
```
Comment (patrickmann) [inline graylog2-server/src/main/java/org/graylog2/indexer/rotation/strategies/TimeBasedSizeOptimizingStrategy.java:113]: `leewayLimit` is now unused.
Can you add a comment briefly explaining the reasoning? It's hard to understand why date are compared this way. In fact I would break it out into a separate method, given that we use it twice. Then it can have a nice descriptive name.

File: graylog2-server/src/main/java/org/graylog2/indexer/rotation/strategies/TimeBasedSizeOptimizingStrategy.java [hunk 4]
```diff
@@ -100,21 +105,25 @@ protected Result shouldRotate(final String index, IndexSet indexSet) {
...
 
     private boolean indexIsOldEnough(DateTime creationDate) {
-        final Instant now = Instant.now();
+        final Instant now = clock.instantNow();
         final Instant rotationLimit = now.minus(ROTATION_PERIOD);
 
-        return creationDate.isBefore(rotationLimit.toEpochMilli());
+        final Duration between = Duration.between(Instant.ofEpochMilli(creationDate.getMillis()), now);
+        return between.compareTo(Duration.ofDays(ROTATION_PERIOD.getDays())) >= 0;
+        //return creationDate.isBefore(rotationLimit.toEpochMilli());
```
Comment (patrickmann) [inline graylog2-server/src/main/java/org/graylog2/indexer/rotation/strategies/TimeBasedSizeOptimizingStrategy.java:122]: `rotationLimit` is now unused

File: graylog2-server/src/test/java/org/graylog2/indexer/rotation/strategies/TimeBasedSizeOptimizingRotationAndRetentionTest.java [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+    }
+
+    private String getBetween(TestIndex i) {
+        if (i.getClosingDate() == null) {
+            return "null";
+        }
+        return org.joda.time.Period.fieldDifference(clock.nowUTC().toLocalDateTime(), i.getClosingDate().toLocalDateTime()).toString();
+    }
+
+    class TestIndexSet extends org.graylog2.indexer.TestIndexSet {
```
Comment (patrickmann) [inline graylog2-server/src/test/java/org/graylog2/indexer/rotation/strategies/TimeBasedSizeOptimizingRotationAndRetentionTest.java:218]: TestIndexSet and TestIndex could be useful for other tests. Maybe break out those into `indexer.indices.util`?

Comment (mpfz0r) [reply graylog2-server/src/test/java/org/graylog2/indexer/rotation/strategies/TimeBasedSizeOptimizingRotationAndRetentionTest.java:218]: hmm, not sure about that. 
it all relies on a lot of mocks (line 107)

**Review by patrickmann — APPROVED** [review]
